### PR TITLE
Implement junit test reporter

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Fixes https://github.com/oven-sh/bun/issues/8019

The biggest issue with this is the lack of stdout/stderr duplicated to the test output.

Example XML:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="bun test" tests="11" assertions="9" failures="4" skipped="2" time="0.029948">
  <testsuite name="test-2.test.js" tests="4" assertions="4" failures="2" skipped="0" time="0" hostname="Jarreds-MacBook-Pro.local">
    <properties>
      <property name="ci" value="https://ci.example.com/123" />
      <property name="commit" value="1234567890" />
    </properties>
    <testcase name="should pass inside nested describe" classname="nested describe" time="0.000907" file="test-2.test.js" assertions="1" />
    <testcase name="should fail inside nested describe" classname="nested describe" time="0.000810375" file="test-2.test.js" assertions="1">
      <failure type="AssertionError" />
    </testcase>
    <testcase name="should pass" classname="" time="0.000162875" file="test-2.test.js" assertions="1" />
    <testcase name="failing test" classname="" time="0.0007653330000000001" file="test-2.test.js" assertions="1">
      <failure type="AssertionError" />
    </testcase>
  </testsuite>
  <testsuite name="passing.test.js" tests="7" assertions="5" failures="2" skipped="2" time="0" hostname="Jarreds-MacBook-Pro.local">
    <properties>
      <property name="ci" value="https://ci.example.com/123" />
      <property name="commit" value="1234567890" />
    </properties>
    <testcase name="should pass inside nested describe" classname="nested describe" time="0.000137458" file="passing.test.js" assertions="1" />
    <testcase name="should fail inside nested describe" classname="nested describe" time="0.000604333" file="passing.test.js" assertions="1">
      <failure type="AssertionError" />
    </testcase>
    <testcase name="should pass" classname="" time="0.000173166" file="passing.test.js" assertions="1" />
    <testcase name="second test" classname="" time="0.000580875" file="passing.test.js" assertions="1" />
    <testcase name="failing test" classname="" time="0.0007428749999999999" file="passing.test.js" assertions="1">
      <failure type="AssertionError" />
    </testcase>
    <testcase name="skipped test" classname="" time="0" file="passing.test.js" assertions="0">
      <skipped />
    </testcase>
    <testcase name="todo test" classname="" time="0" file="passing.test.js" assertions="0">
      <skipped message="TODO" />
    </testcase>
  </testsuite>
</testsuites>

```

### How did you verify your code works?

Tests

### CODE REVIEW

**Comment by robobun** [comment]
:x: @Jarred-Sumner, your commit fe897de4cf6b8c1004cd061229d7073a929ae55c has some failures in [`#6297`](https://buildkite.com/bun/bun/builds/6297)

File: src/cli/test_command.zig [hunk 1]
```diff
@@ -1104,9 +1540,7 @@ pub const TestCommand = struct {
 
         if (reporter.summary.fail > 0 or (coverage.enabled and coverage.fractions.failing and coverage.fail_on_low_coverage)) {
             Global.exit(1);
-        } else if (reporter.jest.unhandled_errors_between_tests > 0) {
-            Global.exit(reporter.jest.unhandled_errors_between_tests);
```
Comment (dylan-conway) [inline src/cli/test_command.zig]: Why was this removed? Was something else supposed to be in the block?

File: src/cli/test_command.zig [hunk 2]
```diff
@@ -76,6 +109,357 @@ fn writeTestStatusLine(comptime status: @Type(.EnumLiteral), writer: anytype) vo
...
+                , .{});
+            },
+            .fail_because_expected_has_assertions => {
+                this.testcases_metrics.failures += 1;
+                try this.contents.writer(bun.default_allocator).print(
+                    \\>
+                    \\      <failure message="Expected to have assertions, but none were run" type="AssertionError"/>
+                    \\    </testcase>
+                , .{});
+            },
```
Comment (dylan-conway) [inline src/cli/test_command.zig:408]: This code is untested. `.fail_because_expected_assertion_count`, `.fail_because_todo_passed`, and `.fail_because_expected_has_assertions` do not appear to be called with `writeTestCase`

File: docs/cli/test.md [hunk 3]
```diff
@@ -55,6 +55,49 @@ $ bun test ./test/specific-file.test.ts
...
+        run: bun install # You can use npm/yarn/pnpm instead if you prefer
+      - name: Run tests
+        run: bun test
+```
+
+From there, you'll get GitHub Actions annotations.
+
+### JUnit XML reports (GitLab, etc.)
+
+To use `bun test` with a JUnit XML reporter, you can use the `--junit-report` flag.
```
Comment (dylan-conway) [inline docs/cli/test.md]:
```suggestion
To use `bun test` with a JUnit XML reporter, you can use the `--reporter=junit` in combination with `--reporter-outfile`.
```

File: docs/cli/test.md [hunk 4]
```diff
@@ -55,6 +55,49 @@ $ bun test ./test/specific-file.test.ts
...
+```
+
+From there, you'll get GitHub Actions annotations.
+
+### JUnit XML reports (GitLab, etc.)
+
+To use `bun test` with a JUnit XML reporter, you can use the `--reporter=junit` in combination with `--reporter-outfile`.
+
+```sh
+$ bun test --junit-report=./bun.xml
```
Comment (dylan-conway) [inline docs/cli/test.md]:
```suggestion
$ bun test --reporter=junit --reporter-outfile=./bun.xml
```

**Review by dylan-conway — APPROVED** [review]

File: src/cli.zig [hunk 5]
```diff
@@ -524,6 +526,23 @@ pub const Arguments = struct {
...
+
+            if (args.option("--reporter")) |reporter| {
+                if (strings.eqlComptime(reporter, "junit")) {
+                    if (ctx.test_options.reporter_outfile == null) {
+                        Output.errGeneric("--reporter=junit expects an output file from --reporter-outfile", .{});
+                        Global.crash();
+                    }
+                    ctx.test_options.file_reporter = .junit;
+                } else {
+                    Output.errGeneric("unrecognized reporter format: '{s}'", .{reporter});
```
Comment (Jarred-Sumner) [inline src/cli.zig]:
```suggestion
                    Output.errGeneric("unrecognized reporter format: '{s}'. Currently, only 'junit' is supported", .{reporter});
```
